text/0000-result_ffi_guarantees.md

```diff
@@ -0,0 +1,80 @@
+# RFC: result_ffi_guarantees
+
+- Feature Name: `result_ffi_guarantees`
+- Start Date: 2023-02-15
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC gives specific layout and ABI guarantees when wrapping "non-zero" data types from `core` in `Option` or `Result`. This allows those data types to be used directly in FFI, in place of the primitive form of the data (eg: `Result<(), NonZeroI32>` instead of `i32`).
```

Comment:
```suggestion
This RFC gives specific layout and ABI guarantees when wrapping "non-zero" data types from `core` in `Result`. This allows those data types to be used directly in FFI, in place of the primitive form of the data (eg: `Result<(), NonZeroI32>` instead of `i32`).
```

Reply:
To the best of my knowledge, having these layout+ABI guarantees on Option hasn't actually been made official by any RFC or similar. The standard library has some examples of where Option wrapping some type will give a same layout, but I'm not sure of anywhere that actually specifies an ABI compatibility.

Reply:
What about https://doc.rust-lang.org/std/option/#representation

Reply:
Re-reading that, it seems to assure memory layout not ABI.

I can remove the mentions of Option stuff if folks think it's absolutely necessary, but it's probably better to have things clear and all in one place. I'm very interested in doing low level and ffi work in Rust, and if even I don't know where to find this info, then we probably haven't written it down enough.

Reply:
`Option`-like enums have their layout and ABI guaranteed by a FCP in https://github.com/rust-lang/rust/pull/60300. Although, this is not currently documented in Rust reference but in [UCG](https://rust-lang.github.io/unsafe-code-guidelines/layout/enums.html) only.